### PR TITLE
[proxima-tools] fix Closure class collection

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: tools/src/test/java/cz/o2/proxima/tools/groovy/ConsoleTest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.stream.Collectors;
+import org.junit.Test;
+
+/** Test {@link cz.o2.proxima.tools.groovy.Console} and various aspects of Closure/classloading. */
+public class ConsoleTest {
+
+  private final Config config = ConfigFactory.load("test-reference.conf").resolve();
+  private final Repository repo = Repository.ofTest(config);
+
+  @Test(timeout = 20000)
```
Comment (dmvk) [inline tools/src/test/java/cz/o2/proxima/tools/groovy/ConsoleTest.java:37]: what takes so long here?

File: tools/src/main/java/cz/o2/proxima/tools/groovy/ToolsClassLoader.java [hunk 2]
```diff
@@ -42,11 +43,16 @@
     }
   }
 
+  final Map<String, byte[]> byteCode = new ConcurrentHashMap<>();
```
Comment (dmvk) [inline tools/src/main/java/cz/o2/proxima/tools/groovy/ToolsClassLoader.java:46]: nit: private

File: tools/src/main/java/cz/o2/proxima/tools/groovy/ToolsClassLoader.java [hunk 3]
```diff
@@ -42,11 +43,16 @@
...
+  final CompilerConfiguration conf = ClassloaderUtils.createConfiguration();
```
Comment (dmvk) [inline tools/src/main/java/cz/o2/proxima/tools/groovy/ToolsClassLoader.java:47]: nit: private

File: tools/src/main/groovy/cz/o2/proxima/tools/groovy/internal/ProximaInterpreter.groovy [hunk 4]
```diff
@@ -35,11 +38,13 @@ public class ProximaInterpreter extends Interpreter {
...
-  public ProximaInterpreter(
+  ProximaInterpreter(
       final ClassLoader classLoader,
-      final Binding binding) {
+      final Binding binding,
+      final CompilerConfiguration configuration) {
 
-    super(classLoader, binding, null);
+    super(classLoader, binding, configuration);
+    Preconditions.checkState(this.shell.getClassLoader() instanceof ToolsClassLoader);
```
Comment (dmvk) [inline tools/src/main/groovy/cz/o2/proxima/tools/groovy/internal/ProximaInterpreter.groovy:47]:
```suggestion
    Preconditions.checkState(getShell.getClassLoader() instanceof ToolsClassLoader);
```

File: tools/src/main/groovy/cz/o2/proxima/tools/groovy/internal/ClassloaderUtils.groovy [hunk 5]
```diff
@@ -34,15 +34,16 @@ import org.codehaus.groovy.syntax.Reduction
...
     ret = ret.addCompilationCustomizers(
         new ASTTransformationCustomizer(TypeChecked),
         new ASTTransformationCustomizer(CompileStatic))
      */
-    ret.setTargetBytecode("1.11")
+    /*
+    ret.setTargetBytecode("11")
     ret.setDebug(true)
-    return ret
+     */
```
Comment (dmvk) [inline tools/src/main/groovy/cz/o2/proxima/tools/groovy/internal/ClassloaderUtils.groovy:46]: should we remove both commented code-blocks? if not, can you please document why?

File: tools/src/main/java/cz/o2/proxima/tools/groovy/Console.java [hunk 6]
```diff
@@ -102,28 +104,33 @@ public static Console create(Config config, Repository repo, String[] args) {
...
   public static void main(String[] args) throws Exception {
-    ClassLoader loader = new ToolsClassLoader();
+    ToolsClassLoader loader = new ToolsClassLoader();
     Thread.currentThread().setContextClassLoader(loader);
-    Console console = Console.get(args);
+    try (Console console = Console.get(args)) {
+      console.run(loader);
+      System.out.println();
+    }
+  }
```
Comment (dmvk) [inline tools/src/main/java/cz/o2/proxima/tools/groovy/Console.java:113]: nice 👍

**Review by dmvk — APPROVED** [review]
LGTM 👍 If I understand correctly, the main difference comes from removing

```
ret.setTargetBytecode("1.11")
```

Maybe can you document, that this needs to be aligned with GroovyClassLoader configuration for the future reference?

Thanks for the patch <3

File: tools/src/main/groovy/cz/o2/proxima/tools/groovy/internal/ProximaInterpreter.groovy (hunk 4 above)
Comment (dmvk) [reply tools/src/main/groovy/cz/o2/proxima/tools/groovy/internal/ProximaInterpreter.groovy:47]: I think this check needs to be removed as it introduces chicken-egg problem :)

- java code needs groovy code to compile
- groovy code needs a java code to compile

File: tools/src/test/java/cz/o2/proxima/tools/groovy/ConsoleTest.java (hunk 1 above)
Comment (je-ik) [reply tools/src/test/java/cz/o2/proxima/tools/groovy/ConsoleTest.java:37]: It generates and compiles the Environment class.

**Pull request merged**